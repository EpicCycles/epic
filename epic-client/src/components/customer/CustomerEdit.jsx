import React from 'react'
import {Dimmer, Loader} from 'semantic-ui-react'
import CustomerDetailEdit from "./CustomerDetailEdit";
import NoteEdit from "../note/NoteEdit";
import {findObjectWithId, updateObject} from "../../helpers/utils";
import CustomerAddressGrid from "./CustomerAddressGrid";
import {
    createEmptyModelWithDefaultFields,
    getModelKey,
    matchesModel
} from "../app/model/helpers/model";
import CustomerPhoneGrid from "./CustomerPhoneGrid";
import {customerNoteFields} from "../app/model/helpers/fields";

class CustomerEdit extends React.Component {
    state = { note: createEmptyModelWithDefaultFields(customerNoteFields)};

    componentDidUpdate(prevProps) {
        if (this.props.notes !== prevProps.notes) {
            const newNoteIsOnList = this.props.notes.some(note => matchesModel(note, customerNoteFields, this.state.note));
            if (newNoteIsOnList) this.setState({ note: createEmptyModelWithDefaultFields(customerNoteFields) })
        }
    }

    saveOrCreateCustomerNote = (note) => {
        if (note.id) {
            this.props.saveNote(note);
        } else {
            const noteToSave = updateObject(note, { customer: this.props.customerId });
            this.props.createNote(noteToSave);
        }
    };

    render() {
        const { note } = this.state;
        const {
            addresses, phones, notes, quotes, customers,
            deleteCustomer, removeCustomer,
            isLoading, customerId,
            deleteNote,
            deleteCustomerPhone, saveCustomerPhone,
            saveCustomerAddress, deleteCustomerAddress,
            saveCustomer, createCustomer
        } = this.props;
        const customer = findObjectWithId(customers, customerId);
        const note_key = getModelKey(note);
        const customer_key = getModelKey(customer);
        return <div id="customer-edit">
            <h2>Customer</h2>
            <section className="row">
                <div>
                    <CustomerDetailEdit
                        customer={customer}
                        saveCustomer={saveCustomer}
                        createCustomer={createCustomer}
                        removeCustomer={removeCustomer}
                        deleteCustomer={deleteCustomer}
                        componentKey={customer_key}
                        key={`detail${customer_key}`}
                        data-test="edit-customer"
                    />
                    {(customerId) &&
                    <div className="grid-container">
                        <CustomerAddressGrid
                            deleteCustomerAddress={deleteCustomerAddress}
                            saveCustomerAddress={saveCustomerAddress}
                            addresses={addresses}
                            customerId={customerId}
                            data-test="edit-customer-addresses"
                        />
                        <CustomerPhoneGrid
                            deleteCustomerPhone={deleteCustomerPhone}
                            saveCustomerPhone={saveCustomerPhone}
                            customerId={customerId}
                            phones={phones}
                            data-test="edit-customer-phones"
                        />
                    </div>}
                </div>
                <div>
                    {(customerId) &&
                    <NoteEdit
                        saveNote={this.saveOrCreateCustomerNote}
                        key={`detail${note_key}`}
                        note={note}
                        deleteNote={deleteNote}
                        data-test="add-customer-note"
                    />}
                </div>
            </section>

            {isLoading &&
            <Dimmer active inverted>
                <Loader content='Loading'/>
            </Dimmer>
            }
        </div>
    }
}

export default CustomerEdit;
