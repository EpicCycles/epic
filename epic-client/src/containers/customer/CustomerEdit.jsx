import React from 'react'
import {Dimmer, Loader} from 'semantic-ui-react'
import ErrorDismissibleBlock from "../../common/ErrorDismissibleBlock";
import CustomerDetailEdit from "./CustomerDetailEdit";
import NoteCreate from "../note/NoteCreate";

class CustomerEdit extends React.Component {
    // add a key to state for each block to force reload on change
    state = {
        note_key: 0,
        customer_key: 1
    };

    saveOrCreateCustomerNote = (note_text, customer_visible) => {
        if (this.props.note && this.props.note.id) {
            let noteToSave = this.props.note;
            noteToSave.note_text = note_text;
            noteToSave.customer_visible = customer_visible;
            this.props.saveNote(noteToSave);
        }
        else {
            const newNote = {
                customer: this.props.customer.id,
                note_text: note_text,
                customer_visible: customer_visible
            };
            this.props.createNote(newNote);
        }
        this.updateNoteKey()
    };

    updateNoteKey = () => {
        if (this.props.note && this.props.note.id) {
            this.setState({note_key: this.props.note.id});
        }
        else {
            this.setState({note_key: 0});
        }
    };

    render() {
        const {removeCustomerError, removeNoteError, acceptCustomerChanges, isLoading, customer, error, note, noteError, removeNote, deleteNote} = this.props;
        const {note_key} = this.state;
        const customer_key = (customer && customer.id) ? customer.id : 0;
        return (
            <div id="customer-edit">
                <h1>Edit Customer</h1>
                {error &&
                <ErrorDismissibleBlock error={error} removeError={removeCustomerError}/>
                }

                <CustomerDetailEdit customer={customer ? customer : {}}
                                    acceptCustomerChanges={acceptCustomerChanges}
                                    key={`detail${customer_key}`}
                />
                {(customer && customer.id) &&
                <NoteCreate saveNote={this.saveOrCreateCustomerNote} removeNoteError={removeNoteError} note={note}
                            key={`detail${note_key}`}
                            noteError={noteError} removeNote={removeNote} deleteNote={deleteNote}
                            updateNoteKey={this.updateNoteKey}
                />}

                {isLoading &&
                <Dimmer active inverted>
                    <Loader content='Loading'/>
                </Dimmer>
                }
            </div>
        )
    }
}

export default CustomerEdit;
