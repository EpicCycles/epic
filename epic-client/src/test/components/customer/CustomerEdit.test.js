import React from 'react';
import CustomerEdit from "../../../components/customer/CustomerEdit";

it("displays a full customer with extras properly", () => {
    const fullCustomer = {
        id:23,
        first_name: "anna",
        last_name: "Blogs",
        email: "anna@blogs.co.uk",
        addresses: ["1", "2"],
        phones: ["10", "20"],
    };
    const note = {
        id: 278,
        note_text: "a note text",
        customer_visible: true,
        customer: fullCustomer.id
    };
    const component = shallow(
        <CustomerEdit customer={fullCustomer} note={note}/>
    );
    expect(component).toMatchSnapshot();
});
it("displays a new customer (no id) properly", () => {
    const newCustomer = {
        first_name: "anna",
        last_name: "Blogs",
        email: "anna@blogs.co.uk",
        addresses: ["1", "2"],
        phones: ["10", "20"],
    };
    const note={};
    const component = shallow(
        <CustomerEdit customer={newCustomer} note={note}/>
    );
    expect(component).toMatchSnapshot();
});
it("calls create customer when a new customer is saved", () => {
    const saveCustomer = jest.fn();
    const createCustomer = jest.fn();

    const fullCustomer = {
        first_name: "anna",
        last_name: "Blogs",
        email: "anna@blogs.co.uk",
        addresses: ["1", "2"],
        phones: ["10", "20"],
    };
    const note = {
        note_text: "a note text",
        customer_visible: true,
    };
    const component = shallow(
        <CustomerEdit
            customer={fullCustomer}
            note={note}
            createCustomer={createCustomer}
            saveCustomer={saveCustomer}
        />
    );
    component.instance().saveOrCreateCustomer("new first", "new last","email@provider.com");
    expect(saveCustomer.mock.calls.length).toBe(0);
    expect(createCustomer.mock.calls.length).toBe(1);
    expect(createCustomer.mock.calls[0][0].id).toBe(undefined);
    expect(createCustomer.mock.calls[0][0].first_name).toBe("new first");
    expect(createCustomer.mock.calls[0][0].last_name).toBe("new last");
    expect(createCustomer.mock.calls[0][0].email).toBe("email@provider.com");
});
it("calls save customer when an existing customer is saved", () => {
    const saveCustomer = jest.fn();
    const createCustomer = jest.fn();

    const fullCustomer = {
        id:23,
        first_name: "anna",
        last_name: "Blogs",
        email: "anna@blogs.co.uk",
        addresses: ["1", "2"],
        phones: ["10", "20"],
    };
    const note = {
        id: 278,
        note_text: "a note text",
        customer_visible: true,
        customer: fullCustomer.id
    };
    const component = shallow(
        <CustomerEdit
            customer={fullCustomer}
            note={note}
            createCustomer={createCustomer}
            saveCustomer={saveCustomer}
        />
    );
    component.instance().saveOrCreateCustomer("new first", "new last");
    expect(createCustomer.mock.calls.length).toBe(0);
    expect(saveCustomer.mock.calls.length).toBe(1);
    expect(saveCustomer.mock.calls[0][0].id).toBe(fullCustomer.id);
    expect(saveCustomer.mock.calls[0][0].first_name).toBe("new first");
    expect(saveCustomer.mock.calls[0][0].last_name).toBe("new last");
    expect(saveCustomer.mock.calls[0][0].email).toBe(undefined);
});
it("calls create note when a new note is saved", () => {
    const saveCustomer = jest.fn();
    const createCustomer = jest.fn();
    const saveNote = jest.fn();
    const createNote = jest.fn();

    const fullCustomer = {
        id: 234,
        first_name: "anna",
        last_name: "Blogs",
        email: "anna@blogs.co.uk",
        addresses: ["1", "2"],
        phones: ["10", "20"],
    };
    const note = {};
    const component = shallow(
        <CustomerEdit
            customer={fullCustomer}
            note={note}
            createCustomer={createCustomer}
            saveCustomer={saveCustomer}
            saveNote={saveNote}
            createNote={createNote}
        />
    );
    component.instance().saveOrCreateCustomerNote("new note ext", false);
    expect(saveCustomer.mock.calls.length).toBe(0);
    expect(createCustomer.mock.calls.length).toBe(0);
    expect(saveNote.mock.calls.length).toBe(0);
    expect(createNote.mock.calls.length).toBe(1);
    expect(createNote.mock.calls[0][0].id).toBe(undefined);
    expect(createNote.mock.calls[0][0].note_text).toBe("new note ext");
    expect(createNote.mock.calls[0][0].customer_visible).toBe(false);
    expect(createNote.mock.calls[0][0].customer).toBe(fullCustomer.id);
});

it("calls save note when an existing note is saved", () => {
    const saveCustomer = jest.fn();
    const createCustomer = jest.fn();
   const saveNote = jest.fn();
    const createNote = jest.fn();

    const fullCustomer = {
        id: 234,
        first_name: "anna",
        last_name: "Blogs",
        email: "anna@blogs.co.uk",
        addresses: ["1", "2"],
        phones: ["10", "20"],
    };
    const note = {
        id: 278,
        note_text: "a note text",
        customer_visible: true,
        customer: fullCustomer.id
    };
    const component = shallow(
        <CustomerEdit
            customer={fullCustomer}
            note={note}
            createCustomer={createCustomer}
            saveCustomer={saveCustomer}
            saveNote={saveNote}
            createNote={createNote}
        />
    );
    component.instance().saveOrCreateCustomerNote("new note ext", false);
    expect(saveCustomer.mock.calls.length).toBe(0);
    expect(createCustomer.mock.calls.length).toBe(0);
    expect(createNote.mock.calls.length).toBe(0);
    expect(saveNote.mock.calls.length).toBe(1);
    expect(saveNote.mock.calls[0][0].id).toBe(note.id);
    expect(saveNote.mock.calls[0][0].note_text).toBe("new note ext");
    expect(saveNote.mock.calls[0][0].customer_visible).toBe(false);
    expect(saveNote.mock.calls[0][0].customer).toBe(note.customer);
});
