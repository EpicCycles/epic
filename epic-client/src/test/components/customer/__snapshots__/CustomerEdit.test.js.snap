// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a full customer with extras properly 1`] = `
<div
  id="customer-edit"
>
  <h2>
    Customer
  </h2>
  <section
    className="row"
  >
    <div>
      <CustomerDetailEdit
        acceptCustomerChanges={[Function]}
        customer={
          Object {
            "addresses": Array [
              "1",
              "2",
            ],
            "email": "anna@blogs.co.uk",
            "first_name": "anna",
            "id": 23,
            "last_name": "Blogs",
            "phones": Array [
              "10",
              "20",
            ],
          }
        }
        key="detail23"
      />
      <React.Fragment>
        <h3>
          Customer Addresses
        </h3>
        <table>
          <tbody>
            <CustomerAddressEdit
              customerId={23}
              key="editNewAddressnew"
            />
            <CustomerAddressEdit
              customerAddress="1"
              customerId={23}
              key="editAddressundefined"
            />
            <CustomerAddressEdit
              customerAddress="2"
              customerId={23}
              key="editAddressundefined"
            />
          </tbody>
        </table>
        <h3>
          Customer Phone
        </h3>
        <table>
          <tbody>
            <CustomerPhoneEdit
              customerId={23}
              key="editNewPhonenew"
            />
            <CustomerPhoneEdit
              customerId={23}
              customerPhone="10"
              key="editPhoneundefined"
            />
            <CustomerPhoneEdit
              customerId={23}
              customerPhone="20"
              key="editPhoneundefined"
            />
          </tbody>
        </table>
      </React.Fragment>
    </div>
    <div>
      <NoteCreate
        key="detail278"
        note={
          Object {
            "customer": 23,
            "customer_visible": true,
            "id": 278,
            "note_text": "a note text",
          }
        }
        saveNote={[Function]}
      />
    </div>
  </section>
</div>
`;

exports[`displays a new customer (no id) properly 1`] = `
<div
  id="customer-edit"
>
  <h2>
    Customer
  </h2>
  <section
    className="row"
  >
    <div>
      <CustomerDetailEdit
        acceptCustomerChanges={[Function]}
        customer={
          Object {
            "addresses": Array [
              "1",
              "2",
            ],
            "email": "anna@blogs.co.uk",
            "first_name": "anna",
            "last_name": "Blogs",
            "phones": Array [
              "10",
              "20",
            ],
          }
        }
        key="detail0"
      />
    </div>
    <div />
  </section>
</div>
`;
